mos-platform/geos-cbm/geos_constants.h

```diff
@@ -0,0 +1,752 @@
+#define KRNL_IO_IN 0x36     /** kernal and I/O mapped into memory */
+#define KRNL_BAS_IO_IN 0x37 /** kernal, I/O and basic mapped into memory */
+
+/** values for config - C128 mmu */
+
+#define CIOIN 0x7E        /** 60K RAM, 4K I/O space in */
+#define CRAM64K 0x7F      /** 64K RAM */
+#define CKRNLBASIOIN 0x40 /** kernal, I/O and basic ROM's mapped into memory   \
+                           */
+#define CKRNLIOIN 0x4E    /** Kernal ROM and I/O space mapped in */
+
+/***********************************************************************
+ * Menu Equates
+ ***********************************************************************/
+
+/** Menu types */
+typedef enum __attribute__((__packed__)) {
+  HORIZONTAL = 0b00000000,
+  VERTICAL = 0b10000000,
+  CONSTRAINED = 0b01000000,
+  UN_CONSTRAINED = 0b00000000
+} menu_type_t;
+
+/** Offsets to variables in the menu structure */
+#define OFF_MY_TOP 0      /** offset to y pos of top of menu */
+#define OFF_MY_BOT 1      /** offset to y pos of bottom of menu */
+#define OFF_MX_LEFT 2     /** offset to x pos of left side of menu */
+#define OFF_MX_RIGHT 4    /** offset to x pos of right side of menu */
+#define OFF_NUM_M_ITEMS 6 /** offset to Alignment|Movement|Number of items */
+#define OFF_1ST_M_ITEM 7  /** offset to record for 1st menu item in structure  \
+                           */
+
+/** Types of menu actions */
+typedef enum __attribute__((__packed__)) {
+  MENU_ACTION = 0x00,  /** for setting byte in menu table that indicates */
+  DYN_SUB_MENU = 0x40, /** whether the menu item causes action */
+  SUB_MENU = 0x80      /** or sub menu */
+} menu_action_t;
+
+/***********************************************************************
+ * Process Related Equates
+ ***********************************************************************/
+
+/** Possible values for processFlags */
+
+#define SET_RUNABLE 0b10000000 /** runable flag */
+#define SET_BLOCKED 0b01000000 /** process blocked flag */
+#define SET_FROZEN 0b00100000  /** process frozen flag */
+#define SET_NOTIMER 0b00010000 /** not a timed process flag */
+
+#define RUNABLE_BIT 7 /** runable flag */
+#define BLOCKED_BIT 6 /** process blocked flag */
+#define FROZEN_BIT 5  /** process frozen flag */
+#define NOTIMER_BIT 4 /** not a timed process flag */
+
+/***********************************************************************
+ * Text Equates
+ ***********************************************************************/
+
+/** Bit flags in mode */
+
+#define SET_UNDERLINE 0b10000000
+#define SET_BOLD 0b01000000
+#define SET_REVERSE 0b00100000
+#define SET_ITALIC 0b00010000
+#define SET_OUTLINE 0b00001000
+#define SET_SUPERSCRIPT 0b00000100
+#define SET_SUBSCRIPT 0b00000010
+#define SET_PLAINTEXT 0b00000000
+
+#define UNDERLINE_BIT 7
+#define BOLD_BIT 6
+#define REVERSE_BIT 5
+#define ITALIC_BIT 4
+#define OUTLINE_BIT 3
+#define SUPERSCRIPT_BIT 2
+#define SUBSCRIPT_BIT 1
+
+/** Putchar constants */
+
+#define EOF 0       /** end of text object */
```

Comment:
`EOF` and `NULL` are defined elsewhere in the C standard library; do they need to be defined here?
It's also the case that many of the below have standard ASCII character escapes (e.g., BS is `\b`). Do we need all these?

Again, if this is intending to be backwards compatbile with something else, that's another question. NULL and EOF may still cause issues; in particular, NULL is usually pointer typed, and EOF is usually negative one.

Reply:
These definitions are from the GEOS original documentation. As no one should include stdio etc I would leave the definitions in here.

Reply:
I think EOF is probably fine, but NULL is also defined in `stdlib.h` and `string.h`, and both are likely to be useful on GEOS (right?). As-is, one will get a strange duplicate warning error for the two definitions, and a failure on `-Werror`.

Presumably the GEOS documents were for an assembly language interface, not C, right? The canonical way to spell a null character in C is `\0`, and that does also appear to be the correct character here. C NULL is also explicitly intended to be used as null pointer constant, not a null character constant.

This is a place where some negotiation needs to happen between C standard practice and legacy practices. If we do need these constants, special casing them to GEOS_EOF and GEOS_NULL seems warranted.

Reply:
I grepped through common/include. None of these files defines NULL...

Maybe they include some standard clang header which do?

I would really like to have NULL defined in GEOS headers, as its used at several places (see also the *.c in example/geos-cbm)...

Reply:
NULL is defined in the `stddef.h` chain that comes directly from the clang compiler. Defining it is considered part of the responsibilities of the compiler. In particular, in C++ it's defined as a compiler intrinsic `__null`. (I just learned this today by looking at the source.)

> I would really like to have NULL defined in GEOS headers, as its used at several places (see also the *.c in example/geos-cbm)...

I couldn't see any references in the example files. Is there a technical issue preventing use of `\0` in the contexts where NULL is to be used?

Reply:
Removed NULL, one can include stddef.h if needed